RFC [WIP] an intermediate representation for SQL

knz opened the pull request:
Reaction to last week's meeting (with @irfansharif @arjunravinarayan @petermattis @RaduBerinde @andreimatei)

@eisenstatdavid I'd like you to help me shape this up before we submit it for discussion to a larger group. Some special questions for you: 
- choice of HLL, would something else read better to the uneducated reader?
- how much more than this do we need to be able to define tree transformations in the HLL and have this translated to Go automatically?
- do you recall the algorithm to check whether a pattern match is exhaustive, or a reference to such an algorithm?

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10055)

<!-- Reviewable:end -->

eisenstatdavid commented:
1. OCaml is fine.
2. Not much more? Rather than define a toy language and write the transformations operationally though, maybe we should implement a term rewriting system and have the compiler write the traversals for us.
3. Determining whether a pattern match is exhaustive is coNP-hard via the obvious reduction from CNF-SAT. Seems like people use this in practice, which also helps compiling the pattern matches efficiently: http://stackoverflow.com/q/7883023 .

---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KUSPEjgqkUyZAB9rn2r:b-3kwaoc)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for your input. Do you know of anything close to term rewriting systems / libraries available in Go already?

andreimatei commented:
I'd love more discussion about why we need to write the IR definition in ML, versus writing it in Go.
Sounds like you're not suggesting we actually write transformations in ML, are you? Then what's the advantage of using this other language and generating Go structs? What's the code that'd be hard to write in Go? Is it the serialization? Or the matchers? (for the matchers, I commented elsewhere that I think we can write a generic one). 

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql-ir.md, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUuie_fc3pDK7O_9I:-KUSUuie_fc3pDK7O_9J:b1hbnqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L15)):*
> ```Markdown
> # Motivation
> 
> Primary motivations:
> ```

how about an extra one - to allow the distsql logical planner to recognize subtrees that it can compile? Or more generally, to allow different compilation "backends".

---

*[docs/RFCS/sql-ir.md, line 17 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUNOXMhmy-RudMAMe:-KUSUNOXMhmy-RudMAMf:b-n073r3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L17)):*
> ```Markdown
> Primary motivations:
> 
> - to encode the query of non-materialized views.
> ```

can you please expand this bullet?

---

*[docs/RFCS/sql-ir.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUdlZTQgdssTLS-bX:-KUSUdlZTQgdssTLS-bY:b4q27hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L19)):*
> ```Markdown
> - to encode the query of non-materialized views.
> - to enable more straightforward definitions for query optimization algorithms.
> - to transfer filter expressions across the wire when spawning distSQL processors.
> ```

is this one a legitimate need? I think we can transfer the filter expressions just fine today, as the string from the current Expression Node

---

*[docs/RFCS/sql-ir.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSVhlM4GQEFqWQ7weO:-KUSVhlM4GQEFqWQ7weP:b59pf0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L23)):*
> ```Markdown
> Side concerns that influence the design:
> 
> - robustness against function renames across cockroachdb versions.
> ```

I'm not sure I follow these constraints... Is this assuming that we're gonna save IR to use later/on a different node? Or is this about saving views?

---

*[docs/RFCS/sql-ir.md, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSXjJ4O2aOR_yUsHKb:-KUSXjJ5lgpNIexaxFJ1:b-6x7nv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L117)):*
> ```Markdown
> ```go
> oexpr := match(iexpr, 
> 	"BinOp(Gt, a, b)",              func(a, b Expr) Expr { return BinOp{Lt, b, a} },
> ```

on the left hand side you could also have structures (struct literals), not strings. 

---

*[docs/RFCS/sql-ir.md, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSeFrSTg37kBz7bFlv:-KUSeFrT5EjzlG9arkNT:b-s9kfwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L128)):*
> ```Markdown
> var matchers := map[string]matcher
> 
> func match1(iexpr Expr, actionfn func(...interface{})) {
> ```

I'm not sure why/if we need to  generate these specialized matchers/"actionfns". Can't we write a generic matching algorithm?
Is the concern performance? Cause I'd argue we should be then first be working on prepared statements and caching this stuff.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KUSfxic6nn1_Elm6ZM6:belic9g)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I also don't quite understand why we need to start with a new language. I thought that the idea is to start with the current `parser` nodes and add more as necessary.

In particular, I don't see the benefit for expressions. We already have all the machinery to use, analyze, normalize (and even a simple way to serialize/deserialize) SQL expressions, why maintain two different languages/grammars/etc? It would mean that whenever we want to add some feature and support something we would need to mess with both.

In the IR, how do you express that you are using a certain index? The IR needs to allow representing this right? (In fact, I can't even find how the general idea of reading from a table is represented, IOW what IR "node" would correspond to the current `scanNode`? Looks like `tref` is only used for insert/update/delete)

knz commented:
Guys I have completed the 2nd iteration, the text is rather .. different now. PTAL.

Note it's probably not complete yet, we likely want to be a bit more specific as to what we want as attributes in the IR nodes in this text (i.e. before they are implemented), but I'd like your opinion about the overall shape of the work before we dive into specifics.

cc @cuongdo @arjunravinarayan @irfansharif @petermattis @nvanbenschoten

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 429 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWREGgN-gQ-BDXfzgzf:-KWREGgN-gQ-BDXfzgzg:b1xpuwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L429)):*
> ```Markdown
>    - regular join, with annotations:
>      - join predicate: cross/on/using/natural/equality (also see issue #10630)
>  	 - which algorithm to use
> ```

This doc mixes tabs and spaces in a few places; I don't know what the markdown renderer will do with that. Replace all tabs with spaces. 

---

*[docs/RFCS/sql-ir.md, line 476 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWRE_u3WVL9nefquoQx:-KWRE_u3WVL9nefquoQy:b92pqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L476)):*
> ```Markdown
> 1. The IR nodes must be appropriately *named*, at least in documentation,
>    so as to provide a common language to drive further discussions and serve
>    as references in code TODOs. (FIXME: perhaps this RFC should do just this?)
> ```

Yeah, I think the RFC is the appropriate place to define names. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Great stuff, thanks for all the additions! Overall shape/direction is great!

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 437 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZgpDQo-gCMdieyjcH:-KWZgpDQo-gCMdieyjcI:bhq9ama) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L437)):*
> ```Markdown
>    - show
>    - explain
>    - index join (this is really a hybrid)
> ```

Note that the "join by lookup" method could be used for other joins, not  just index join. I agree it's somewhat of a hybrid though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWZiYIAMee1rF-D7mNP:b-cwibky)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

What I'd like a bit more detail on is how exactly the "backends" are going to interact with the IR tree. Are we going to pass the IR to DistSQL, which then has an opportunity to do more transformations? And then DistSQL is going to delegate the parts that it doesn't want to deal with to another "backend" that does further transformations and then produces planNodes?

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSXjJ4O2aOR_yUsHKb:-KWZsNupn3cJKRtdVOEv:b-anmuc4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L117)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> on the left hand side you could also have structures (struct literals), not strings. 
</details>
I still maintain my comment

---

*[docs/RFCS/sql-ir.md, line 14 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZdlDnfwhMMrmn3YxX:-KWZdlDnfwhMMrmn3YxY:bwprd0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L14)):*
> ```Markdown
> 
> Summarized implementation strategy:
> 1. identify transforms (we're not doing great on this yet)
> ```

nit: this doesn't render on new lines for some reason. Maybe you need an empty line above `1.`

---

*[docs/RFCS/sql-ir.md, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZlATlVhnt6GT0J9Sc:-KWZlATlVhnt6GT0J9Sd:b3prpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L29)):*
> ```Markdown
> - to encode the query of non-materialized views.
> - to provide robustness against db/table/column/view renames in a lightweight fashion.
> - (hopefully) to version built-in function and virtual table references.
> ```

There's multiple references to this versioning throughout the document, but it's not very clear (at least to me) what the point is. Might be worth explaining a bit. Or dropping.
Is it about using prepared statements /views between versions of the server? If so, would we really maintain multiple versions of descriptors/functions, as opposed to "recompiling" everything for every new version?

---

*[docs/RFCS/sql-ir.md, line 380 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZkEDdvY1oBO803LcT:-KWZkEDdvY1oBO803LcU:bycswl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L380)):*
> ```Markdown
> 	  - strength reduction (e.g. `a LIKE 'foo%'` to `a >= "foo" AND a < "fop"`)
> 	  - comparison simplification
> 	  - **[new]** normalization to disjunctive normal form (an OR of ANDs)
> ```

you might want to clarify that this helps with index selection

---

*[docs/RFCS/sql-ir.md, line 381 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZk7B69mzCzu5ay3Ny:-KWZk7B69mzCzu5ay3Nz:b-17bkjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L381)):*
> ```Markdown
> 	  - comparison simplification
> 	  - **[new]** normalization to disjunctive normal form (an OR of ANDs)
> 	  - partial evaluation of disjunctions and conjunctions
> ```

what does this mean?

---

*[docs/RFCS/sql-ir.md, line 402 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZkVyW6RZzk2LlipN5:-KWZkVyW6RZzk2LlipN6:bc92w3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L402)):*
> ```Markdown
>    2. **[new]** ordering propagation (this can simplify nodes) 
> 
>    3. **[new]** ordering relaxation, where every ordering that is not
> ```

can you give an example of this?

---

*[docs/RFCS/sql-ir.md, line 551 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZrY0V5yZBaG7f-14X:-KWZrY0WNaRfn9zxUCQJ:bgkrl85) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L551)):*
> ```Markdown
> - SQL's "IN" has two IR encodings, depending on whether one of the operands is a subquery.
> 
> The bulk of what makes SQL special is the IR for "table expressions".
> ```

"table expressions" are "2. logical nodes that act as data sources" from your taxonomy above?

---

*[docs/RFCS/sql-ir.md, line 557 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZs0SRtxjlx7R01-nV:-KWZs0SRtxjlx7R01-nW:batqu83) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L557)):*
> ```Markdown
> - the IR for table expressions reifies the operators from relational
>   algebra: Join Filter, Render, Sort, etc.
> - Insert, Update, Delete, Explain and Show are also table expressions in the IR.
> ```

for `INSERT/UPDATE/DELETE`, you might want to explain that a `COUNT` operation gets magically placed above the table expression if `RETURNING` is not used.

---

*[docs/RFCS/sql-ir.md, line 639 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZsoiQocbVs1QmU_cW:-KWZsoiQocbVs1QmU_cX:b-9phrl8) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L639)):*
> ```Markdown
> 
> - Expression tree can contain nodes with Go values (expecially
>   timestamps, dates). What to do with those with respect to
> ```

@raduberinde what does DistSQL do with them today?

---

*[docs/RFCS/sql-ir.md, line 914 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZwOJkHVmdjWrK6T0B:-KWZwOJkHVmdjWrK6T0C:b-2hc0vj) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L914)):*
> ```Markdown
> (* DoAndCount(T) means run the clause T and return the count of its results. *)
> (* For example INSERT/UPDATE/DELETE is initially encoded as DoAndCount(<Clause>). *)
> (* DoAndCount(T) then disappears during normalization into Do(GroupAgg(T, *, COUNT( * ))). *)
> ```

I wonder how we'll actually run these `DoAndCounts`... Seems to me like, with any backend, we wouldn't actually put a `COUNT` node, but instead we'd directly configure the inserter to return the count. And maybe DistSQL would put a summer processor on top. How would that fit? Would there be a further "normalization" step, possibly dependent of the backend, that would remove this `COUNT`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWZx65Rsg-6uZX12A-C:b-y2prrh)*
<!-- Sent from Reviewable.io -->

eisenstatdavid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

but is this IR essentially going to serve as the abstract syntax tree too? The phases of the compiler that check for errors (name resolution, type checking) need to operate on something close to the surface syntax to give good error messages, but subsequent phases are simpler if there's a desugaring step that reduces the number of constructs that they must consider.

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWciFD8WplNVh6vTlBC:bexf1du)*
<!-- Sent from Reviewable.io -->

knz commented:
@andreimatei: I have added a paragraph in the summary to explain the interaction.

@eisenstatdavid: I have added a paragraph to clarify that pretty-printing back to SQL will only be possible up to semantic analysis and that we may lose that opportunity after that, and instead we can keep track of the text location that the IR node originates from in the input for error reporting.

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUuie_fc3pDK7O_9I:-KWd2-vucA41Y6v3v1AP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L15)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about an extra one - to allow the distsql logical planner to recognize subtrees that it can compile? Or more generally, to allow different compilation "backends".
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 17 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUNOXMhmy-RudMAMe:-KWd20y2ewUtdoMhS8BD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L17)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you please expand this bullet?
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 19 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSUdlZTQgdssTLS-bX:-KWd22ddn-2SLidHS8uI:b4771y2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L19)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> is this one a legitimate need? I think we can transfer the filter expressions just fine today, as the string from the current Expression Node
</details>
Clarified in text.

---

*[docs/RFCS/sql-ir.md, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSVhlM4GQEFqWQ7weO:-KWd24yTO8rNm6LxBWrX:b-n3mgbi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L23)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'm not sure I follow these constraints... Is this assuming that we're gonna save IR to use later/on a different node? Or is this about saving views?
</details>
We discussed that in a meeting.

---

*[docs/RFCS/sql-ir.md, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSXjJ4O2aOR_yUsHKb:-KWd3ZgPr5Aa9s8We5Tl:biebrbx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L117)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I still maintain my comment
</details>
I think I want to revisit this section after we get some code working.

---

*[docs/RFCS/sql-ir.md, line 14 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZdlDnfwhMMrmn3YxX:-KWd1pzX8M-oc59_9Iss:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L14)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> nit: this doesn't render on new lines for some reason. Maybe you need an empty line above `1.`
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 29 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZlATlVhnt6GT0J9Sc:-KWd27VlljbL2ZHe_4GC:b-z8ta9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L29)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> There's multiple references to this versioning throughout the document, but it's not very clear (at least to me) what the point is. Might be worth explaining a bit. Or dropping.
> Is it about using prepared statements /views between versions of the server? If so, would we really maintain multiple versions of descriptors/functions, as opposed to "recompiling" everything for every new version?
</details>
Clarified in meeting.

---

*[docs/RFCS/sql-ir.md, line 380 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZkEDdvY1oBO803LcT:-KWd2JcoCNyHTuMTinkg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L380)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> you might want to clarify that this helps with index selection
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 381 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZk7B69mzCzu5ay3Ny:-KWd2UsxHFWNQqnauK8e:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L381)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what does this mean?
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 402 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZkVyW6RZzk2LlipN5:-KWd2r-NAiBWaGP5B3vo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L402)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can you give an example of this?
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 429 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWREGgN-gQ-BDXfzgzf:-KWd39uSDJ6wGGfNo61A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L429)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This doc mixes tabs and spaces in a few places; I don't know what the markdown renderer will do with that. Replace all tabs with spaces. 
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 437 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZgpDQo-gCMdieyjcH:-KWd3ECrh_vLPPpvZ7_D:b-eowr3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L437)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Note that the "join by lookup" method could be used for other joins, not  just index join. I agree it's somewhat of a hybrid though.
</details>
Clarified.

---

*[docs/RFCS/sql-ir.md, line 551 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZrY0V5yZBaG7f-14X:-KWd3OaOSba4nQxCRcwI:bo6ap9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L551)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> "table expressions" are "2. logical nodes that act as data sources" from your taxonomy above?
</details>
Indeed!

---

*[docs/RFCS/sql-ir.md, line 557 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZs0SRtxjlx7R01-nV:-KWd3WhofUVWzs1sPUUp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L557)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> for `INSERT/UPDATE/DELETE`, you might want to explain that a `COUNT` operation gets magically placed above the table expression if `RETURNING` is not used.
</details>
Done.

---

*[docs/RFCS/sql-ir.md, line 914 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZwOJkHVmdjWrK6T0B:-KWd3gRnf2WuYHsQSPMc:bw7ab5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L914)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I wonder how we'll actually run these `DoAndCounts`... Seems to me like, with any backend, we wouldn't actually put a `COUNT` node, but instead we'd directly configure the inserter to return the count. And maybe DistSQL would put a summer processor on top. How would that fit? Would there be a further "normalization" step, possibly dependent of the backend, that would remove this `COUNT`?
</details>
I'd hope we can optimize this a bit, since count is distributive and commutative it is easy to parallelize!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWd5fWGXswy06kQk5a2:bfzfiv6)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
@knz it doesn't look like you pushed the changes you made in response to everyone's comments?

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWi5WoixYc99F886zPU:b59rd5g)*
<!-- Sent from Reviewable.io -->

knz commented:
Oh indeed, my bad. Done.

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for writing this all up!

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWj8CO1miNGl3L_Pjc_:-KWj8CO1miNGl3L_Pjca:b-uhdlv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fe87f20a49e706c457355de14fd1749375ecff1/docs/RFCS/sql-ir.md#L51)):*
> ```Markdown
> - The IR needs not be pretty-printable back to valid SQL. In the few
>   cases where something needs to produce SQL syntax back to a user
>   (e.g. SHOW CREATE), the IR node can keep a copy of the string in
> ```

I understand the importance of keeping things simple and completely agree with this decision, but things like showing a view definition to the user after a bunch of the tables/columns it uses have been renamed will unfortunately require more logic than just keeping a copy of the original string.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KWkDgKJupfwtsz6Pxgu:b-kensgo)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWj8CO1miNGl3L_Pjc_:-KWlEg0kAfr58eATz5Uo:bi7zigi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fe87f20a49e706c457355de14fd1749375ecff1/docs/RFCS/sql-ir.md#L51)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> I understand the importance of keeping things simple and completely agree with this decision, but things like showing a view definition to the user after a bunch of the tables/columns it uses have been renamed will unfortunately require more logic than just keeping a copy of the original string.
</details>
Will it? Why not simply store a copy of the IR tree at the end of the semantic check phase in the view descriptor, at a point where it's still pretty-printable? Then complete the IR processing every time the view is loaded (we could possibly introduce a cache)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/sql-ir.md, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWj8CO1miNGl3L_Pjc_:-KWmqmI_OBRyo8GrN23T:b-7pgzu9) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fe87f20a49e706c457355de14fd1749375ecff1/docs/RFCS/sql-ir.md#L51)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Will it? Why not simply store a copy of the IR tree at the end of the semantic check phase in the view descriptor, at a point where it's still pretty-printable? Then complete the IR processing every time the view is loaded (we could possibly introduce a cache)
</details>
That would be great if it's an option. I was just taking the point "The IR needs not be pretty-printable back to valid SQL" at face value :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reminder the discussion in https://github.com/cockroachdb/cockroach/pull/11736#issuecomment-264120729 about "ghost columns partcipating in ordering"

knz committed:
RFC [WIP] an intermediate representation for SQL

knz commented:
I had another iteration on this PR after recent developments:
1. I have a rather simple and elegant solution for the problem "Go data in IR" - will send out separate RFC which will need implementing before this IR merges.
2. I realized thanks to our optimizer training that my view on pattern matching was naive and inadequate. I simply removed that section and pattern matching will probably receive better treatment in a separate RFC.
3. I also realized that the IR for abstract syntax, which describes the user input of a query, is necessarily different from the IR for logical plans. Trying to merge them together as this PR was initially attempting makes the picture murky and hard to read. I will have another iteration to separate them further and probably split this RFC in two.

The big next steps at this point are:
- propose+implement solution to point 1 above
- iterate on preliminary list of IR node names, to complete this RFC
- merge this PR, then continue work in subsequent RFCs.

---

Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions.

---

*[docs/RFCS/sql-ir.md, line 128 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KUSeFrSTg37kBz7bFlv:-KtCfaQxaBY-AeA9L0LJ:bae17gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/4094975cc2b9f27f9a9007c8da4c4aef577d5587/docs/RFCS/sql-ir.md#L128)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm not sure why/if we need to  generate these specialized matchers/"actionfns". Can't we write a generic matching algorithm?
Is the concern performance? Cause I'd argue we should be then first be working on prepared statements and caching this stuff.
</blockquote></details>

Removed this section. Will address in a later / separate RFC.

---

*[docs/RFCS/sql-ir.md, line 476 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWRE_u3WVL9nefquoQx:-KtCfmEkgV9Wy2-J-abK:b-20jjvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L476)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Yeah, I think the RFC is the appropriate place to define names. 
</blockquote></details>

Done.

Will need to iterate a little bit before this merges though.

---

*[docs/RFCS/sql-ir.md, line 639 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWZsoiQocbVs1QmU_cW:-KtCfY6LP6Ff-rkfICH9:bsjhkw7) ([raw file](https://github.com/cockroachdb/cockroach/blob/06718b9948fc4f3d38806626bf6f9196c8bdbd51/docs/RFCS/sql-ir.md#L639)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

@raduberinde what does DistSQL do with them today?
</blockquote></details>

I removed this section - I have a solution, will write a separate RFC.

---

*[docs/RFCS/sql-ir.md, line 51 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KWj8CO1miNGl3L_Pjc_:-KtCbqm1PEAnb_j0dPBo:b8hxa2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6fe87f20a49e706c457355de14fd1749375ecff1/docs/RFCS/sql-ir.md#L51)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

That would be great if it's an option. I was just taking the point "The IR needs not be pretty-printable back to valid SQL" at face value :)
</blockquote></details>

I added a qualifier sub-clause in the sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-KtCfqyZiNaiH_MM_MQy:b45v7yl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql-ir.md, line 389 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KtIRQTIYo9vn40U1zTI:-KtIRQTIYo9vn40U1zTJ:brgk80h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4754887c87de71c107ef3d123c5a365b0801c9f2/docs/RFCS/sql-ir.md#L389)):*
> ```Markdown
>       - etc. (this extracts the logic currently present in planNode constructors)
>       - **at the end of this stage, a statement is valid and can
>         be run**
> ```

If we're going by some of the terminology we've learned recently, I think the above we all be considered "prep".

---

*[docs/RFCS/sql-ir.md, line 530 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-KtIS5WUuMJvK6mdzYik:-KtIS5WUuMJvK6mdzYil:b-c59zlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/4754887c87de71c107ef3d123c5a365b0801c9f2/docs/RFCS/sql-ir.md#L530)):*
> ```Markdown
>    - accessor methods.
>    - walk (visitor) recursion.
>    - match methods meant to ease pattern matching (see below for details)
> ```

I'm still fuzzy on how the pattern matching would ease the writing of transformations. That is, I can imagine this at a high-level, but the specifics are where the devil is. I think it would be worthwhile, perhaps in a companion RFC, to give an example of 2 transformations in the current code and how they would be simplified. Two good examples would be expression normalization and propagating filters across inner joins. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This RFC has been in `draft` form for quite a while. An over-arching concern I have is that it presents justification about one approach while not listing out any alternatives. And I don't have a concrete grasp about whether the proposed IR nodes will be a material improvement for the types of transformations we want to perform during SQL optimization. I think it is worth revisiting this RFC in light of some of the knowledge we've gained recently. 

---

Review status: 0 of 1 files reviewed at latest revision, 21 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/sql-ir.md, line 238 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-Kw1-nqV4WHGX7ALX3gW:-Kw1-nqV4WHGX7ALX3gX:b-n74z9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/4754887c87de71c107ef3d123c5a365b0801c9f2/docs/RFCS/sql-ir.md#L238)):*
> ```Markdown
> sizeable language processing component.
> 
> ## General implementation direction: auto-generate the code
> ```

As my understanding of the needs of this IR have become clearer, I'm not sure if auto-generated code is necessary, at least for the IR node structure. As an alternative, I think a single node structure used for both relational and scalar expressions can satisfy our needs and enable easy transformations. What seems to be important for the IR structure is to be able to easily determine and update logical properties about a node, such as the input and output variables. This could be achieved by making the IR nodes implement an interface (`inputVars(), outputVars(), etc`), or we can pull those logical properties into a `node` structure and have the node-specific data attached via some `hook` field. 

There is still an exploration to be made about whether transformations should be represented in a higher-level language or directly in Go. The latter certainly gives the most flexibility, but perhaps at the expense of expressiveness. 

---

*[docs/RFCS/sql-ir.md, line 361 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-Kw12muHzGBDuYJ85_u2:-Kw12muHzGBDuYJ85_u3:bw7ul7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/4754887c87de71c107ef3d123c5a365b0801c9f2/docs/RFCS/sql-ir.md#L361)):*
> ```Markdown
>    the nodes already defined, so they can be used as drop-in
>    replacement into the existing code (i.e. preserve struct and
>    attribute names)
> ```

This approach to using the IR sounds like it is trying to replace the AST nodes. An alternative would be to leave the AST nodes in place and provide a transformation from the AST to the IR. The advantage of this approach is that we can choose when we perform the transformation. For example, we can leave all of the existing name resolution, type checking and normalization code in place for the time being and use the IR for higher-level transformations and query planning. Over time, we can move the conversion from the AST to the IR closer to the raw AST.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10055#-:-Kw14RwbOmVb9U_Z_sxC:bds49ab)*
<!-- Sent from Reviewable.io -->

knz commented:
Superseded by #19135.
